content/news/XXXX-XX-XX-controller-api-declartions.md

```diff
@@ -0,0 +1,53 @@
+title: Controller API declarations for JavaScript mapping developers
+authors: Jörg Wartenberg
+status: draft
+tags: 2.4, controller
+
+
+If the support of a DJ controller requires more than the semantic 1:1 mapping of MIDI codes, Mixxx offers the possibility to use freely programmable Javascript code to implement such more complex functionalities.
+
+The Mixxx internal Application Programming Interface (API) for such controller scripts was previously only partially documented, in the [Mixxx Wiki](https://github.com/mixxxdj/mixxx/wiki/).
+
+This has changed with Mixxx 2.4, now there are formal TypeScript declarations for each individual API class/method. These can be read by any code editor with [Language Server Protocol (LSP)](https://en.wikipedia.org/wiki/Language_Server_Protocol) support, to be visualized to the developer while typing his controller scripting code.
+![Screenshot of documentation from controller API declaration visualized in editor]({static}/images/news/ControllerApiDocuInEditor.png)
+Popular editors and IDEs with LSP support are Atom, Eclipse, Sublime Text, Visual Studio and Visual Studio Code.
+
+In addition to the inline visualization of the documentation, the syntax is also checked against the TypeScript declarations. Usage errors are reported directly when writing the code in the editor.
+![Screenshot of syntax check against controller API declaration visualized in editor]({static}/images/news/ControllerApiSyntaxCheckInEditor.png)
+
+## How to use it?
+Development of mappings and controller scripts is usually done in the subdirectory `controllers` of the [Mixxx Settings Directory](https://manual.mixxx.org/latest/chapters/appendix/settings_directory), while the common files (like `common-controller-scripts.js`) which comes with Mixxx, are store in the subdirectory `controllers` of Mixxx installation directory, or `./mixxx/res/controllers` if you built Mixxx yourself. This is also the location, where the TypeScript files with the declarations of the Mixxx controller API are stored. The LSP language server need to know the location of all of them. To tell the LSP server general settings and the locations of these files, you need to create a file with the name `jsconfig.json` placed next to your JavaScript controller script file, in the `controllers` directory in your `Mixxx Settings Directory`.
```

Comment:
This is very technical and detailed while not being very relevant timely. 
I think this belongs almost entirely somewhere on the wiki instead.

Reply:
I agree here. Can we describe it in a wiki and just link it here? 
We need a wiki entry anyway. Maybe also a Readme.txt in the controller folder? 
How about installing a jsconfig.json.template file with a comment how to adjust and use it?

Reply:
Sorry, but this is the essence of this blog post! Without the sample content of this file, it is completely incomprehensible for a mapping developer how to use this functionality.
Everything else in this post is just an explanation, but that's what I want to share with the mapping community here!
Feel free to copy elsewhere - but please don't make me remove this post from my blog post.

Reply:
IMHO this certainly deserves to be a newspost. 
This is news, though it's more technical it definitely has it's public.

Articles like this make the difference: Mixxx isn't just a program like 13 in a dozen, there's a lot of added value and although we're not literaly selling something the (shared) knowledge / flexibility / support (not treating users as complete idiots) all are part of Mixxx's USP (Unique Selling Point).
Not everyone reads the wiki on regular bases but rather when they are dealing with a problem or looking for particular information. If articles like this don't get (news-) attention the info might got lost between the other info (that can be overwelming). The readers can choose to read it or not (if it's too technical for them).
The Wiki also needs an 'updated / new' - intro (simular to changelog)

Reply:
> Sorry, but this is the essence of this blog post! Without knowing how to use the sample content of this file, it is completely incomprehensible for a mapping developer how to use this functionality. Everything else in this post is just an explanation, but that's what I want to share with the mapping community here! Feel free to copy elsewhere - but please don't make me remove this post from my blog post.

yeah... makes sense. Lets copy it to an appropriate page in the wiki after publishing the article.

Reply:
Do you have a place in mind, where I should copy it to?

Reply:
maybe right at the beginning of https://github.com/mixxxdj/mixxx/wiki/midi%20scripting#set-up-a-javascript-mapping ?